Added link checking to tox and release.py

steffenschroeder opened the pull request:
Fixes #1722 by checking all links in documentation during the release process.
Should only be merged after all broken links are fixed (https://github.com/pytest-dev/pytest/pull/5613)

It's running during release in release.py or can be triggered manually by executing `tox -e docs-checklinks`.

False positives can be added to list `linkcheck_ignore` in `doc/en/conf.py`.

steffenschroeder commented on scripts/release.py:
```diff
@@ -79,12 +79,19 @@ def fix_formatting():
     regen()
     changelog(version, write_out=True)
     fix_formatting()
+    check_links()
```
Question to the maintainers: 
Should this be mandatory all the time or should there be the possibility to skip this with a command line flag in `release.py`?
Just thinking about a broken/flaky link could prevent an ad-hoc release of pytest (in case of a bad release that needs to be patched quickly)

Zac-HD replied:
IMO link-checking should be a separate thing, where maintainers (or contributors!) can run the script but it doesn't happen in CI - flakiness is inevitable if we automate it.  It should probably go in the "how to release" docs though 👍

steffenschroeder replied:
In https://github.com/pytest-dev/pytest/issues/1722#issuecomment-499690461 @nicoddemus mentioned:

> @mandarvaze just call it from release.py should be fine. 👍

So I followed this. I'm up for any desired change, though.

The-Compiler replied:
@Zac-HD That's already the case - from what I can see `release.py` doesn't get run on CI, but gets run when cutting a release (via `tox -e release`).

Zac-HD replied:
Alright!  No change needed then :smile: 

Thanks @The-Compiler for clarifying and @steffenschroeder for your work on this :heart:

steffenschroeder commented on doc/en/conf.py:
```diff
@@ -112,6 +112,16 @@
+    "https://github.com/numpy/numpy/blob/master/doc/release/1.16.0-notes.rst#new-deprecations",
+    "https://blogs.msdn.microsoft.com/bharry/2017/06/28/testing-in-a-cloud-delivery-cadence/",
+    "http://pythontesting.net/framework/pytest-introduction/",
+]
```
this link can be used to maintain false-positives

blueyed replied:
Can those be regular expressions?
(Might be worth adding a link to its doc there - and sorry for being too lazy to search for it myself)

steffenschroeder replied:
They can be regular expressions. I updated the comment accordingly. 
Thanks for the suggestion.

blueyed replied:
Thanks.

Why is "https://github.com/numpy/numpy/blob/master/doc/release/1.16.0-notes.rst#new-deprecations" ignored?

I also suggest to ignore `https://github.com/pytest-dev/pytest/issues/\d+` (a lot of links, and those should be OK in general)

blueyed replied:
Also:

- linkcheck_workers=20 is a bit much maybe (the default is 5, and probably good).

Seen this failure:
```
(line   41) broken    http://pacha.cafepais.com/ - HTTPConnectionPool(host='pacha.cafepais.com', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7ff9cc55a438>: Failed to establish a new connection: [Errno -2] Name or service not known'))
```

And I am surprised to see that it does not use an internal cache?!  (or does it just print the same result again)

```
writing output... [ 97%] usage

(line  707) redirect  http://bpaste.net - permanently to https://bpaste.net/
(line  367) redirect  http://docs.python.org/library/pdb.html - permanently to https://docs.python.org/3/library/pdb.html
(line  370) redirect  http://docs.python.org/library/pdb.html - permanently to https://docs.python.org/3/library/pdb.html
(line  332) redirect  http://docs.python.org/library/pdb.html - permanently to https://docs.python.org/3/library/pdb.html
(line  336) redirect  http://docs.python.org/library/pdb.html - permanently to https://docs.python.org/3/library/pdb.html
(line  336) redirect  http://docs.python.org/library/pdb.html - permanently to https://docs.python.org/3/library/pdb.html
```

Anyway, I suggest the following patch:

- given that it is a builtin module, the doc reference link is not really needed - sorry for not checking this before
- I think it should not be quiet, since it takes very long, and it is good to see progress

```diff
diff --git i/doc/en/conf.py w/doc/en/conf.py
index 93fbc4787..5c78bca38 100644
--- i/doc/en/conf.py
+++ w/doc/en/conf.py
@@ -112,16 +112,18 @@
 # A list of ignored prefixes for module index sorting.
 # modindex_common_prefix = []
 
-# A list of regular expressions that match URIs that should not be checked when doing a lin>
-# see https://www.sphinx-doc.org/en/master/usage/configuration.html#options-for-the-linkche>
+# A list of regular expressions that match URIs that should not be checked when
+# doing a linkcheck.
 linkcheck_ignore = [
     "https://github.com/numpy/numpy/blob/master/doc/release/1.16.0-notes.rst#new-deprecatio>
     "https://blogs.msdn.microsoft.com/bharry/2017/06/28/testing-in-a-cloud-delivery-cadence>
     "http://pythontesting.net/framework/pytest-introduction/",
+    r"https://github.com/pytest-dev/pytest/issues/\d+"
+    r"https://github.com/pytest-dev/pytest/pull/\d+"
 ]
 
-# The number of worker threads to use when checking links
-linkcheck_workers = 20
+# The number of worker threads to use when checking links (default=5).
+# linkcheck_workers = 5
 
 
 # -- Options for HTML output ---------------------------------------------------
diff --git i/tox.ini w/tox.ini
index 37577bdae..6a3c9b888 100644
--- i/tox.ini
+++ w/tox.ini
@@ -72,7 +72,7 @@ usedevelop = True
 changedir = doc/en
 deps = -r{toxinidir}/doc/en/requirements.txt
 commands =
-    sphinx-build -q -W --keep-going -b linkcheck . _build
+    sphinx-build -W --keep-going -b linkcheck . _build
 
 [testenv:doctesting]
 basepython = python3
```

blueyed replied:
Let me know if I should just push it - I'd also squash it then already.

But there are still some 404s anyway - I've read somewhere that there's another PR/plan to fix them?

steffenschroeder replied:
Thanks for the suggestions.
Since you already spend the efforts I‘d be happy if you pushed this.
If it more convenient if I apply the patch, I’ll do this tonight

The-Compiler replied:
Why ignore pytest GitHub links? Seems useful to know when you accidentally typo'd an issue/PR number and end up with an issue which doesn't exist.

blueyed replied:
@The-Compiler 
Because there are a lot, and it slows the whole process down.
To check if they exist, you could basically just check if they are below the current maximum also (since issues redirect to pulls anyway).
Typos might happen, but it is usually for some other (existing) issue then.

blueyed replied:
Still wonder about:

> Why is "https://github.com/numpy/numpy/blob/master/doc/release/1.16.0-notes.rst#new-deprecations" ignored?

blueyed replied:
> Since you already spend the efforts I‘d be happy if you pushed this.

Can do so later (have it in a stash), if we all agree on it.
Apart from that I'd like to have it squashed into your commits anyway.

steffenschroeder replied:
I applied and pushed the patch.

> Why is "https://github.com/numpy/numpy/blob/master/doc/release/1.16.0-notes.rst#new-deprecations" ignored?

`linkcheck` treats this as _broken_, although it's working fine. Hence I added it to the list of false-positives.

steffenschroeder replied:
Personally, be removing `-q` it find the output very noise, but this might be a matter of personal taste.

steffenschroeder replied:
> And I am surprised to see that it does not use an internal cache?! (or does it just print the same result again)

`linkcheck` does not appear to do any caching here. Might be worth to propose an upstream PR later

blueyed replied:
> Personally, be removing `-q` it find the output very noise, but this might be a matter of personal taste.

I got the impression that it would hang.. and since it runs for minutes.. but well.

blueyed commented on doc/en/conf.py:
```diff
@@ -112,16 +112,18 @@
 # A list of ignored prefixes for module index sorting.
```
Please do not use mentions in commit messages.

steffenschroeder replied:
sorry for that - will the gone one the commit is squash-merged.

blueyed replied:
You need to force-push then, squash-merging is not enabled here.. :/

blueyed replied:
..but you can use `[ci skip]` then in the first line of the commit message then probably.

steffenschroeder replied:
Any objections against me squahing all commits onto one and force pushing that one?

blueyed replied:
No.. :)

steffenschroeder replied:
Done. Is there anything else missing from your point of view?

nicoddemus commented:
Hi,

First of all thanks @steffenschroeder for the work on this! We appreciate it.

I've just tried this branch and there are still link check errors. It takes a while, but the worst part is that it doesn't really show a summary of the broken links, forcing me to scroll and hunt for the missing ones on the terminal. Am I missing something?

Executing `tox -e docs-checklinks | grep broken`, I get:

```
(line   44) broken    http://pytest.org/2.0.0/assert.html#newreport - Anchor 'newreport' not found
(line   10) broken    http://pytest.org/2.0.0/index.html - 404 Client Error: Not Found for url: http://pytest.org/2.0.0/index.html
(line   39) broken    http://pytest.org/2.0.0/usage.html - 404 Client Error: Not Found for url: http://pytest.org/2.0.0/usage.html
(line   53) broken    http://pytest.org/2.0.0/customize.html - 404 Client Error: Not Found for url: http://pytest.org/2.0.0/customize.html
(line   59) broken    http://pytest.org/latest/plugins.html#cmdunregister - Anchor 'cmdunregister' not found
(line   77) broken    http://pytest.org/latest/mark.html - 404 Client Error: Not Found for url: http://pytest.org/latest/mark.html
(line   21) broken    http://pytest.org/latest/example/markers.html - 404 Client Error: Not Found for url: http://pytest.org/latest/example/markers.html
(line   12) broken    http://pytest.org/latest/example/parametrize.html - 404 Client Error: Not Found for url: http://pytest.org/latest/example/parametrize.html
(line   30) broken    http://pytest.org/latest/funcarg_compare.html - 404 Client Error: Not Found for url: http://pytest.org/latest/funcarg_compare.html
(line   36) broken    http://pytest.org/latest/getting-started.html - 404 Client Error: Not Found for url: http://pytest.org/latest/getting-started.html
(line   21) broken    http://pytest.org/latest/unittest.html - 404 Client Error: Not Found for url: http://pytest.org/latest/unittest.html
(line   16) broken    http://pytest.org/latest/fixture.html - 404 Client Error: Not Found for url: http://pytest.org/latest/fixture.html
(line   97) broken    http://pytest.org/latest/faq.html - 404 Client Error: Not Found for url: http://pytest.org/latest/faq.html
(line   16) broken    http://pytest.org/latest/example/parametrize.html - 404 Client Error: Not Found for url: http://pytest.org/latest/example/parametrize.html
(line    9) broken    http://pytest.org/latest/yieldfixture.html - 404 Client Error: Not Found for url: http://pytest.org/latest/yieldfixture.html
(line   54) broken    https://pytest.org/latest/contributing.html - 404 Client Error: Not Found for url: https://pytest.org/latest/contributing.html
(line  126) broken    https://pytest.org/latest/usage.html#modifying-python-traceback-printing - Anchor 'modifying-python-traceback-printing' not found
(line 3124) broken    https://docs.pytest.org/en/latest/builtin.html#pytest.approx - Anchor 'pytest.approx' not found
(line 4476) broken    https://pytest.org/latest/usage.html#modifying-python-traceback-printing - Anchor 'modifying-python-traceback-printing' not found
(line 5000) broken    https://pytest.org/latest/contributing.html - 404 Client Error: Not Found for url: https://pytest.org/latest/contributing.html
(line 5739) broken    http://pytest.org/latest/example/parametrize.html - 404 Client Error: Not Found for url: http://pytest.org/latest/example/parametrize.html
(line 5875) broken    http://pytest.org/latest/faq.html - 404 Client Error: Not Found for url: http://pytest.org/latest/faq.html
(line 5984) broken    http://pytest.org/latest/mark.html - 404 Client Error: Not Found for url: http://pytest.org/latest/mark.html
(line 6170) broken    http://pytest.org/plugins.html#cmdunregister - Anchor 'cmdunregister' not found
(line   31) broken    http://code.activestate.com/pypm/ - HTTPConnectionPool(host='code.activestate.com', port=80): Max retries exceeded with url: /pypm/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001B02609D908>: Failed to establish a new connection: [WinError 10060] Uma tentativa de conex▒o falhou porque o componente conectado n▒o respondeu\r\ncorretamente ap▒s um per▒odo de tempo ou a conex▒o estabelecida falhou\r\nporque o host conectado n▒o respondeu',))
```

A lot of those seem to me to come from old CHANGELOG entries or the CHANGELOG itself, which I'm not sure how valuable it is in practice to fix links on the entry for 2.0.0.

An idea: would it be possible to only check links in the actual documentation? Those definitely should be up-to-date.

steffenschroeder commented:
The verbose output (showing working links) can be suppressed by adding ‘-q’ to the call of the sphinx-build. I had that initially but removed per per previous suggestion.
Can add it again.
The builder has also an option to disable the check of html anchors.
Unfortunately it has no option to exclude files (like old change log).
So could come up with some own link checker, enhance the sphinx linkchecker.
Alternatively we could continue checking everything.

The-Compiler commented:
FWIW I'd prefer fixing the links everywhere (or removing them if they aren't relevant anymore)

nicoddemus commented:
> The verbose output (showing working links) can be suppressed by adding ‘-q’ to the call of the sphinx-build. I had that initially but removed per per previous suggestion.
Can add it again.

Indeed: https://github.com/pytest-dev/pytest/pull/5614#discussion_r305562108. Personally I prefer `-q` because then we only see the failing links, instead of a ton of text which then forces you to "hunt" for the broken links. @blueyed?

Also we need volunteers to fix the broken links before we can merge this.

steffenschroeder commented:
I already fixed a couple of broken links with another PR.
If there are still some broken ones left, I can fix them (once I‘m back from vacation in 2 weeks).

steffenschroeder commented:
Removed the verbose output (only broken links are shown) and fixed the remaining broken links.

steffenschroeder commented:
Hi,
this has unfortunately lost some traction.(I acknowledge that this is not high priority though)

I fixed again all broken links and updated this PR. 

For me, I see the following options to continue:
1. This PR get merged as it is. (Link checking mandatory during release process)
2. We make the check for the broken optional (maybe through a command line flat to release.py)
3. We decouple link checking from release process (leave it in tox but remove it from release.py) to let maintainers or volunteers do the checking from time to time.
4. We remove the link checking completely and only keep the already fixed broken links.
5. We close the PR without further action. 
6. something completely different :-)

All of the above is fine for me and if there is some follow up, I'm happy to work and that. 
But I think we should decide on the next step and close this one.

blueyed commented:
Can you rebase it to have green CI at least for now?
(might have done it but learnt that some folks do not like it)

nicoddemus commented:
@steffenschroeder thanks again for your patience on this! 👍 

I've tested the branch just now and `tox -e docs-checklinks` took ~4mins and completed successfully, which I believe it is acceptable time when cutting a release, and broken links should be far and between.

I propose we merge this (after rebasing as suggested by @blueyed) and see how it fits on the release process, as it is not something we can't rethink later.

steffenschroeder commented:
Thanks for considering this to merge.
I rebased the branch.

Zac-HD commented:
Close-and-reopening to restart CI...

Zac-HD commented:
Thanks again @steffenschroeder!  This is a really useful change which will help the many people who read our docs, and I really admire your persistence in seeing it through ❤️

blueyed commented:
Thanks indeed!

Just for reference, since I've tried it:

> 31.80s user 1.70s system 17% cpu 3:14.54 total
